### PR TITLE
DM-30242: Environment-specific documentation deployments

### PR DESCRIPTION
**Pull request by jonathansick**
This PR implements Science Platform environment-specific documentation deployments (content is customized as needed for each science platform).

The approach taken to writing environment-specific documentation is outlined in these pages:

- https://nb.lsst.io/v/DM-30242/project/environment-specific-documentation.html
- https://nb.lsst.io/v/DM-30242/project/rsp-environments-yaml.html

I've used these techniques to customize the content in:

- `index.rst` ➡️  https://nb.lsst.io/v/DM-30242/index.html
- `getting-started/account.rst` ➡️  https://nb.lsst.io/v/DM-30242/getting-started/account.html
- `getting-started/logging-in.rst` ➡️ https://nb.lsst.io/v/DM-30242/getting-started/logging-in.html

At the moment, I'm not doing ticket branch documentation deployments for each environment (it's a possibility, but I'd want to redesign the edition dashboard pages first). The `/v/` ticket branch builds correspond to the `idfprod` environment. To see documentation builds for the other environments, you can download their sites as artifacts from GitHub Actions e.g. https://github.com/lsst-dm/nb_lsst_io/actions/runs/937326905

In implementing this work, we're also modernizing the build infrastructure

- Adopt tox for managing builds
- Use pip-tools to manage dependencies for reproducible builds
- Use GitHub Actions for CI/CD

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      matrix:
+        rsp:
+          - base
+          - idfdev
+          - idfint
+          - idfprod
+          - int
+          - nts
+          - stable
+          - summit
```
Comment (SimonKrughoff) [inline .github/workflows/ci.yaml:41]: Do you want to include the Tucson test stand if it ever reappears?

File: conf.py [hunk 2]
```diff
@@ -1,201 +1,174 @@
...
+# rsp-environments.yaml is keyed by environment tag name and contains
+# metadata about each RSP environment.
+env_yaml = Path(__file__).parent.joinpath("rsp-environments.yaml").read_text()
+rsp_envs = yaml.safe_load(env_yaml)
+env_names = rsp_envs.keys()
+# Select the environment given the sphinx tag (-t on sphinx-build CLI)
+# Default to idfprod if a tag is not set.
+rsp_env = None
+for env_name in rsp_envs.keys():
+    if env_name in tags:  # noqa: F821
```
Comment (SimonKrughoff) [inline conf.py:33]: I'm sure this is the reason for the `noqa`, but how is `tags` populated?  I don't see it assigned anywhere.

File: requirements/dev.in [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+#
+# After editing, update requirements/dev.txt by running:
+#     make update-deps
+
+-c main.txt
+
+mypy
+types-PyYAML
+requests
+types-requests
```
Comment (SimonKrughoff) [inline requirements/dev.in:12]: Why is tox installed by the action rather than being a proper requirement?

File: src/getting-started/account.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
+##################
+Getting an account
+##################
+
+.. jinja:: rsp
+
+   {% if phalanx == "idfprod" %}
+   You need an account to use the |rsp-at|.
+   Accounts are available to DP0.1 delegates.
+   To learn more about DP0.1 and how to become a delegate and receive a Rubin Science Platform account, visit the `DP0.1 Guide`_.
```
Comment (SimonKrughoff) [inline src/getting-started/account.rst:10]: Where is the link for the DP0.1 Guide defined?

File: src/getting-started/logging-in.ldfvpnsteps.in.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+Log into the Notebook Aspect
+----------------------------
+
+1. |open-nb-aspect| in your web browser or click on the Notebook Aspect link from the |rsp| homepage.
+
+2. Click the **Sign in with CILogon** button.
+
+3. On the new page, choose **NCSA** as the provider unless you have explicitly set up another identity provider (type ``NCSA`` in the search box to find it).
+
+4. Click the **Log On** button
```
Comment (SimonKrughoff) [inline src/getting-started/logging-in.ldfvpnsteps.in.rst]: All of the other items have a period at the end.  I think for consistency this one should too

File: src/getting-started/logging-in.rst [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+   {% else %}
+   Before you get started, make sure you have an account.
+   See :doc:`account` for details.
+   {% endif %}
+
+.. jinja:: rsp
+
+   {% if phalanx in ("stable", "int") %}
+   .. include:: logging-in.ldfvpnsteps.in.rst
+   {% elif phalanx in ("idfint", "idfprod") %}
```
Comment (SimonKrughoff) [inline src/getting-started/logging-in.rst:21]: What about `idfdev`?

File: src/getting-started/logging-in.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+-----
+
+The machine images are based on `LSST Science Pipelines`_ Docker images, which are built on the CentOS 7 Linux operating system.
+
+You can choose which version of the LSST Science Pipelines to run:
+
+- Recommended
+- Release (``rX.Y``)
+- Weekly releases (``YYYY_WW``)
+- Daily releases (``YYYY_MM_DD``)
```
Comment (SimonKrughoff) [inline src/getting-started/logging-in.rst:44]: The weekly and daily are prefixed by `w_` and `d_` respectively.  I'm not sure if it was intentional to leave that off

File: src/project/contributing.rst [hunk 8]
```diff
@@ -43,34 +46,72 @@ These are the basic steps to clone and build the docs:
    git clone https://github.com/lsst-dm/nb_lsst_io
    cd nb_lsst_io
 
-Next, create a Python virtual environment (with `venv <https://docs.python.org/3/tutorial/venv.html>`__, for example).
+Next, create a Python virtual environment (with venv_, for example).
```
Comment (SimonKrughoff) [inline src/project/contributing.rst:49]: Here's another example where a URL must be defined, but I'm not sure how

Comment (SimonKrughoff) [reply src/project/contributing.rst:49]: Oh, maybe it's a link to an anchor in this doc.  If so, ignore me.

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+setenv =
+    base: RSPDOC_ENV = base
+    idfdev: RSPDOC_ENV = idfdev
+    idfint: RSPDOC_ENV = idfint
+    idfprod: RSPDOC_ENV = idfprod
+    int: RSPDOC_ENV = int
+    nts: RSPDOC_ENV = nts
+    stable: RSPDOC_ENV = stable
+    summit: RSPDOC_ENV = summit
+    tucsonteststand: RSPDOC_ENV = tucsonteststand
```
Comment (SimonKrughoff) [inline tox.ini:20]: As I noted before, this seems to be included in some places, but not others

File: src/project/environment-specific-documentation.rst [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+Information about the RSP environments is maintained in a file named ``rsp-environments.yaml`` in the `documentation repository <https://github.com/lsst-dm/nb_lsst_io>`_.
+To learn more about this data, how it is maintained, and how to add additional metadata, see :doc:`rsp-environments-yaml`.
+
+.. _envdocs-substitutions:
+
+Using reStructuredText substitutions
+====================================
+
+For inline content (words or sentences) that changes from one environment to another, such as the name of the science platform environment or a link to the science platform, use `reStructuredText substitutions <https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions>`__.
+You can find these substitutions defined inside the ``rst_epilog`` inside the ``conf.py`` file at the root of the ``nb_lsst_io`` repository, and particularly within the ``rst_epilog.jinja`` templated file.
```
Comment (SimonKrughoff) [inline src/project/environment-specific-documentation.rst]: I think there needs to be a comma after rst_epilog

File: src/project/environment-specific-documentation.rst [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+      {% if primary %}
+      This paragraph appears in the documentation for the
+      primary science platform environment.
+
+      {% elif phalanx in ("base", "summit") %}
+      This paragraph appears in documentation for the base
+      and summit.
+
+      {% phalanx == "stable" %}
```
Comment (SimonKrughoff) [inline src/project/environment-specific-documentation.rst]: Does this not need an `elif`?

File: src/project/environment-specific-documentation.rst [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+      This paragraph appears only for the LDF "stable"
+      environment.
+
+      {% endif %}
+
+The argument to the ``jinja`` directive is always ``rsp``.
+This setting makes all of the fields for a specific RSP environment from the :doc:`rsp-environments.yaml file <rsp-environments-yaml>` available in the Jinja environment.
+The example above uses the ``primary`` and ``phalanx`` fields (the ``phalanx`` field refers to the name of the environment in Phalanx).
```
Comment (SimonKrughoff) [inline src/project/environment-specific-documentation.rst]: Perhaps link to phalanx?

File: src/project/environment-specific-documentation.rst [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+When using the ``jinja`` directive, as with any Sphinx directive, ensure that content is indented consistently with respect to the scope of the directive, as shown above.
+
+.. _envdocs-jinja-includes:
+
+Using Jinja templating with source file includes (\*.in.rst)
+============================================================
+
+The previous approach works well for templating paragraphs, however it is inconvenient to write inside a Jinja directive (within the scope of Jinja syntax, at that.
```
Comment (SimonKrughoff) [inline src/project/environment-specific-documentation.rst]: Missing closing paren

**Review by SimonKrughoff — APPROVED** [review]
This looks really great.  I only have minor comments.

File: conf.py (hunk 2 above)
Comment (jonathansick) [reply conf.py:33]: It just exists in the Sphinx config -> https://www.sphinx-doc.org/en/master/usage/configuration.html?highlight=tags

Comment (SimonKrughoff) [reply conf.py:33]: OK

File: requirements/dev.in (hunk 3 above)
Comment (jonathansick) [reply requirements/dev.in:12]: Tox runs outside your environment (it actually creates standalone virtual environments) so we don't include it with the requirements. Same for pre-commit.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (jonathansick) [reply .github/workflows/ci.yaml:41]: I was thinking of holding off on this until the Tucson test stand became a thing.

File: src/getting-started/account.rst (hunk 4 above)
Comment (jonathansick) [reply src/getting-started/account.rst:10]: In `conf.py` so that it's available to all pages.

**Comment by jonathansick** [comment]
Thanks for the edits @SimonKrughoff. At this point, I'm learning towards keeping all the infrastructure in the nb_lsst_io repo to handle `tucsonteststand`, but leave it out of the GitHub Actions matrix because we don't have an immediate use for that documentation deployment.

**Pull request merged**